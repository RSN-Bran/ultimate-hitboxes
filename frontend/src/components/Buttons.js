import React from "react"
import ToolTip from './ToolTip';

import minus from '../media/minus.png'
import plus from '../media/plus.png'
import play from '../media/play.png'
import pause from '../media/pause.png'
import next from '../media/next.png'
import previous from '../media/previous.png'

import '../css/Button.css';

function Buttons(props) {
	console.log(props.index)
	return (
		<div>

			<img
				data-tip data-for="previousToolTip"
				className={props.index !== 0 ? "button" : "buttonNoClick"}
				id="previous"
				src={previous}
				onClick={props.index !== 0 ? props.previousMove : null}
				alt="Previous Move"
			/>

			<ToolTip
				id="previousToolTip"
				text="Show Previous Move"
				render={props.index !== 0}
			/>

			<img
				data-tip data-for="minusToolTip"
				className={props.currentFrame !== 1 && !props.playing ? "button" : "buttonNoClick"}
				id="minus"
				src={minus}
				onClick={props.currentFrame !== 1 && !props.playing ? props.decrementFrame : null}
				alt="Decrement Frame"
			/>

			<ToolTip
				id="minusToolTip"
				text="Go Back 1 Frame"
				render={props.currentFrame !== 1 && !props.playing}
			/>

			<img
				data-tip data-for="playToolTip"
				className={props.totalFrames !== 1 ? "button" : "buttonNoClick"}
				id="pause-play"
				src={props.playing ? pause : play}
				onClick={props.playing ? props.pause : props.play}
				alt="Play Move"
			/>
			<ToolTip
				id="playToolTip"
				text="Play the Move"
				render={props.totalFrames !== 1}
			/>

			<img
				data-tip data-for="plusToolTip"
				className={props.currentFrame !== props.totalFrames && !props.playing ? "button" : "buttonNoClick"}
				id="plus"
				src={plus}
				onClick={props.currentFrame !== props.totalFrames && !props.playing ? props.incrementFrame : null}
				alt="Increment Frame"
			/>
			<ToolTip
				id="plusToolTip"
				text="Go Forward 1 Frame"
				render={props.currentFrame !== props.totalFrames && !props.playing}
			/>

			<img
				data-tip data-for="nextToolTip"
				className={props.index !== props.totalMoves - 1 ? "button" : "buttonNoClick"}
				id="next"
				src={next}
				onClick={props.index !== props.totalMoves - 1 ? props.nextMove : null}
				alt="NextMove"
			/>
			<ToolTip
				id="nextToolTip"
				text="Show Next Movee"
				render={props.index !== props.totalMoves - 1}
			/>

		</div>
	)
}

export default Buttons