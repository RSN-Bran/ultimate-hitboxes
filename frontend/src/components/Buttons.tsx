//React Imports
import * as React from "react"
import {Link} from 'react-router-dom'

//Component Imports
import ToolTip from './ToolTip';

//CSS Imports
import '../css/Button.css';

//Import Media
import IMAGES from '../media/media_imports.js'

//Find the next available move by moving forwards or backwards until a move that is set to completed is found
function getNextPrev(index, characterData, increment) {
	let move=undefined
	index = index+increment
	while(index>=0 && index < characterData.moves.length) {
		if(characterData.moves[index].completed) {
			move=characterData.moves[index].value
			break
		}
		else{
			index = index+increment
		}
	}
	return move

}
//Render 5 buttons for manipulating the video. Each button has a corresponding tool tip
//Previous: Go the the previous move in the move list, not selectable if the video is currently playing or if there is no available previous move
//Decrement: Move the video back by one frame, not selectable if the move if the video is currently playing or the move is on frame 1
//Play/Pause: If the video is playing, shows a pause button to pause the video. If the video is puased, shows a play button to play the 
//Increment: Move the video forward by one frame, not selectable if the move if the video is currently playing or the move is on its last frame
//Next: Go the the next move in the move list, not selectable if the video is currently playing or if there is no available next move
function Buttons(props) {

	try { 
		//Establish the index of the move within the move list
		let index = props.currentCharacterData.moves.findIndex(move => move.value === props.currentMoveData.value)

		//Set the move to be passed as the next move in the list
		let nextMove=getNextPrev(index, props.currentCharacterData, 1)
		let prevMove=getNextPrev(index, props.currentCharacterData, -1)

		return (
			<div id="buttons">

				<Link to={prevMove !== undefined ? `/${props.currentCharacterData.value}/${prevMove}` : null}>
					<img
						data-tip data-for="previousToolTip"
						className={index !== 0 ? "button" : "buttonNoClick"}
						id="previous"
						src={IMAGES[`previous_${props.settings.theme}`]}
						onClick={() => { props.newMove(prevMove) }}
						alt="Previous Move"
					/>
				</Link>
				<ToolTip
					id="previousToolTip"
					text="Show Previous Move"
					render={index !== 0}
				/>

				<img
					data-tip data-for="minusToolTip"
					className={props.currentFrame !== 1 && !props.playing ? "button" : "buttonNoClick"}
					id="minus"
					src={IMAGES[`minus_${props.settings.theme}`]}
					onClick={() => { props.currentFrame !== 1 && !props.playing ? props.setCurrentFrame(props.currentFrame - 1) : null }}
					alt="Decrement Frame"
				/>

				<ToolTip
					id="minusToolTip"
					text="Go Back 1 Frame"
					render={props.currentFrame !== 1 && !props.playing}
				/>

				<img
					data-tip data-for="playToolTip"
					className={props.currentMoveData.frames !== 1 ? "button" : "buttonNoClick"}
					id="pause-play"
					src={props.playing ? IMAGES[`pause_${props.settings.theme}`] : IMAGES[`play_${props.settings.theme}`]}
					onClick={() => {props.setPlaying(!props.playing)}}
					alt="Play Move"
				/>
				<ToolTip
					id="playToolTip"
					text={props.playing ? "Pause the Move" : "Play the Move"}
					render={props.currentMoveData.frames !== 1}
				/>

				<img
					data-tip data-for="plusToolTip"
					className={props.currentFrame !== props.currentMoveData.frames && !props.playing ? "button" : "buttonNoClick"}
					id="plus"
					src={IMAGES[`plus_${props.settings.theme}`]}
					onClick={() => { props.currentFrame !== props.currentMoveData.frames && !props.playing ? props.setCurrentFrame(props.currentFrame + 1) : null }}
					alt="Increment Frame"
				/>
				<ToolTip
					id="plusToolTip"
					text="Go Forward 1 Frame"
					render={props.currentFrame !== props.currentMoveData.frames && !props.playing}
				/>

				<Link to={nextMove !== undefined ? `/${props.currentCharacterData.value}/${nextMove}` : null}>
					<img
						data-tip data-for="nextToolTip"
						className={index !== props.currentCharacterData.moves.length - 1 ? "button" : "buttonNoClick"}
						id="next"
						src={IMAGES[`next_${props.settings.theme}`]}
						onClick={() => { props.newMove(nextMove) }}
						alt="NextMove"
					/>
				</Link>
				<ToolTip
					id="nextToolTip"
					text="Show Next Move"
					render={index !== props.totalMoves - 1}
				/>
			</div>
		)
	}
	catch (err) {
		return null
		
	}
}

export default Buttons