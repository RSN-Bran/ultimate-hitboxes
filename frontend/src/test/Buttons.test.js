import React from 'react'
//import {render} from "@testing-library/react"
import {create, act} from 'react-test-renderer';
import {shallow} from 'enzyme'

import Buttons from '../components/Buttons'

import "@testing-library/jest-dom/extend-expect"
import { Router } from 'react-router-dom'

import * as mockCharacterData from './mockData/mock-characterdata.json'
import * as mockMoveData from './mockData/mock-movedata.json'

describe ('Buttons component', () => {
    it('Buttons exist', () => {
        const wrapper = shallow(<Buttons
             currentCharacterData={mockCharacterData}
             currentMoveData={mockMoveData}
             settings={{theme:"dark"}}
             playing={true}
            />)
    })

})

test('Slider Snapshot Test', () => {

    let func = (frame) => {
        console.log(frame)
    }

    let app;
    act(()=>{
        app = create(<Buttons/>)
    })


    expect(app.toJSON()).toMatchSnapshot()
    
})