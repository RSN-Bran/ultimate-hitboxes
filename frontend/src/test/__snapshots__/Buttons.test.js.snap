// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider Snapshot Test 1`] = `null`;
